examples/directml/stable_diffusion_xl/stable_diffusion_xl.py

```diff
@@ -0,0 +1,369 @@
+                    "path": Path(optimized_olive_model.model_path),
+                },
+            }
+
+            print(f"Unoptimized Model : {model_info[submodel_name]['unoptimized']['path']}")
+            print(f"Optimized Model   : {model_info[submodel_name]['optimized']['path']}")
+
+    # Save the unoptimized models in a directory structure that the diffusers library can load and run.
+    # This is optional, and the optimized models can be used directly in a custom pipeline if desired.
+    print("\nCreating ONNX pipeline...")
+
+    if getattr(pipeline, "feature_extractor", None) is not None:
+        feature_extractor = pipeline.feature_extractor
+    else:
+        feature_extractor = None
+
+    # TODO: Enable inference with XL models when available
+    # XL model inference is not supported yet because the diffusers library doesn't have an ONNX pipeline for it
+    onnx_pipeline = ORTStableDiffusionXLPipeline(
+        vae_encoder_session=OnnxRuntimeModel.load_model(
+            model_info["vae_encoder"]["unoptimized"]["path"].parent / "model.onnx"
+        ),
+        vae_decoder_session=OnnxRuntimeModel.load_model(
+            model_info["vae_decoder"]["unoptimized"]["path"].parent / "model.onnx"
+        ),
+        text_encoder_session=OnnxRuntimeModel.load_model(
+            model_info["text_encoder"]["unoptimized"]["path"].parent / "model.onnx"
+        ),
+        unet_session=OnnxRuntimeModel.load_model(model_info["unet"]["unoptimized"]["path"].parent / "model.onnx"),
+        text_encoder_2_session=OnnxRuntimeModel.load_model(
+            model_info["text_encoder_2"]["unoptimized"]["path"].parent / "model.onnx"
+        ),
+        tokenizer=pipeline.tokenizer,
+        tokenizer_2=pipeline.tokenizer_2,
+        scheduler=pipeline.scheduler,
+        feature_extractor=feature_extractor,
+        config=dict(pipeline.config),
+    )
+
+    print("Saving unoptimized models...")
+    onnx_pipeline.save_pretrained(unoptimized_model_dir)
+
+    # Create a copy of the unoptimized model directory, then overwrite with optimized models from the olive cache.
+    print("Copying optimized models...")
+    shutil.copytree(unoptimized_model_dir, optimized_model_dir, ignore=shutil.ignore_patterns("weights.pb"))
+    for submodel_name in submodel_names:
+        src_path = model_info[submodel_name]["optimized"]["path"]
+        dst_path = optimized_model_dir / submodel_name / "model.onnx"
+        shutil.copyfile(src_path, dst_path)
+
+        weights_src_path = src_path.parent / (src_path.name + ".data")
+        if weights_src_path.is_file():
+            weights_dst_path = dst_path.parent / (dst_path.name + ".data")
+            shutil.copyfile(weights_src_path, weights_dst_path)
+
+    print(f"The optimized pipeline is located here: {optimized_model_dir}")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_id", default="stabilityai/stable-diffusion-xl-base-1.0", type=str)
```

Comment:
Seems like most of this script is identical to the SD 1.x/2.x example. I'm not as familiar with the XL version, but would it be easy to simply extend the other sample based on the model_id? It looks like you might need to change the static dim names and perhaps some of the JSON. I could have missed something. Do you anticipate it being easier to have these scripts separated?

Reply:
I anticipate having them separated will make it easier. For once, they use different pipelines and the types that the pipelines take as arguments are different.

As I enable inference for the refiner model, this will make the logic even more complex by nature (refiner works by first running the base XL model, taking the result and then running refiner on top of it), as well as having user scripts that are going to diverge.

I think the refiner model should also eventually move to the XL example, which should simplify the logic of the vanilla SD example.